User fingerprints

WithoutPants opened the pull request:
Resolves #99 

- Added new `user_id` column to `scene_fingerprints`. The migration creates an internal user `_legacy_submissions` and assigns the existing fingerprints to this user. All fingerprints will have their `submissions` counter reset to 1 as a result. This migration requires the `pgcrypto` extension to be installed to generate the created user id. 
- Added `user_ids` to `FingerprintInput`. This allows multiple users to be assigned to a single fingerprint when mutating scene fingerprints. If it is empty, or if the user does not have the modify role, then this is set to the current user only.
- Adds `unmatch` flag to `FingerprintSubmission`. When true, unmatches the fingerprint to the scene for the listed or current user
- `submissions` and `updated` has been deprecated and is now unused in `FingerprintEditInput`. Submissions is now derived from the number of user-hash relationships, and updated is derived this way as well.
- moved the user authorisation code into the `user` package
- refactored the scene integration tests (and some infrastructure code) to use a graphql client instead of poking into the resolvers manually
- moved scene mutation code into a new `scene` package

When getting scene fingerprints, the `submissions` field is resolved from the count of the scene/algorithm/hash group. `duration` is resolved as the average of the `duration` fields for the group. This is not ideal, but I'm not sure how else to handle it. `created` and `updated` are based on the `min` and `max` creation dates.

The update operation has been changed to maintain existing UI behaviour. This means that it removes any fingerprints that aren't in the provided input, maintains existing fingerprints that are provided (without changing the associated users), and adds any new fingerprints provided (setting the user id to the current user if not provided).

WithoutPants committed:
WIP

WithoutPants committed:
Set users on fingerprint if omitted

WithoutPants committed:
Fix authorization bug

WithoutPants committed:
Refactor integration tests to use graphql client

WithoutPants committed:
Integration tests for submitFingerprint

WithoutPants committed:
Reorganise code

WithoutPants committed:
Handle scene update fingerprints

WithoutPants committed:
Update integration tests

WithoutPants committed:
Add extension to initdb script

WithoutPants committed:
Set current user id if not provided during update

WithoutPants committed:
Add build tag

WithoutPants committed:
Linting

WithoutPants committed:
Add dependencies

WithoutPants committed:
Merge remote-tracking branch 'upstream/develop' into user-fingerprints

WithoutPants committed:
Remove fingerprints from scene edits

WithoutPants committed:
Remove fingerprints from scene edit in UI

WithoutPants committed:
Change fingerprint duration to mode

WithoutPants commented:
As part of this change, I've removed the ability to edit scene fingerprints with scene edits. Scene fingerprints should be manipulated using the `submitFingerprint` instead. 

I have also changed the output of `duration` to be the `mode` of the submissions, rather than the average. This means that the most popular duration value of the submissions for a hash will be shown.

ghost commented:
To support stash I think we need a field on Fingerprints to denote whether a user has submitted it or not. That way we can display that in stash and prompt the user to remove it if necessary. It would also make it a lot easier to manage fingerprint submissions after a scene has been tagged.

I also think it would be useful to have a `Submissions` field on fingerprints that detail the individual submissions, but that can wait for a separate PR.

WithoutPants committed:
Merge remote-tracking branch 'upstream/develop' into user-fingerprints

WithoutPants committed:
Add user_submitted to Fingerprint

WithoutPants committed:
Add ui-fmt make target

WithoutPants committed:
Show user submitted fingerprints

WithoutPants commented:
> To support stash I think we need a field on Fingerprints to denote whether a user has submitted it or not. That way we can display that in stash and prompt the user to remove it if necessary. It would also make it a lot easier to manage fingerprint submissions after a scene has been tagged.

Done. I added a `user_submitted` field on `Fingerprint`. It required an extra query to match the user submitted fingerprints with the aggregate groups. 

I also changed the fingerprints view to show if the current user has submitted a fingerprint:

![image](https://user-images.githubusercontent.com/53250216/138085374-d624d04f-51b9-45dc-8f6c-4b98465344cb.png)

> I also think it would be useful to have a `Submissions` field on fingerprints that detail the individual submissions, but that can wait for a separate PR.

There is already a numeric `submissions` field on `Fingerprint`. I think I need more context as to what this would be used for.

WithoutPants committed:
Fix unit test compile errors

Commit:
Merge branch 'develop' into user-fingerprints

Commit:
Update migration number and remove dead code

Commit:
Optimize fingerprint fetching and fix compilation errors

Commit:
Remove unused function
